File: python/206.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        h = ListNode(None)
        p = head
        while p:
            rest = p.next
            p.next = h.next
            h.next = p
            p = rest
        return h.next

File: cpp/206.cpp
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
  ListNode *reverseList(ListNode *head, ListNode* tail) {
    if(head == nullptr)
      return tail;
    else{
      ListNode *rest = head->next;
      head->next = tail;
      return reverseList(rest, head);
    }
  }

public:
  ListNode *reverseList(ListNode *head) {
    return reverseList(head, nullptr);
  }
};
